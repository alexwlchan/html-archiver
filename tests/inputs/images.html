<html>

<head>
  <meta charset="utf-8">
  <title>This is a page with images</title>
</head>

<body>
  <p>This is a page to test how html-archiver handles images.</p>

  <p>Because <code>&gt;img&lt;</code> tags can refer to external resources, they should be inlined where possible.</p>

  <!-- A basic image that works -->
  <img src="rainbow.png"/>

  <!-- We can include it more than once -->
  <img src="rainbow.png"/>

  <!-- An image that is missing -->
  <img src="missing.jpg">

  <!-- Test that other attributes are preserved -->
  <img alt="a green snake with a red tongue" src="snake.gif" style="border: 1px solid black;">

  <!-- Test that URL encoded junk is interpreted correctly -->
  <img src="spiral.jpeg?color=black&amp;alt_color=white">

  <!-- Test that if an image is already base64-encoded, it's ignored -->
  <img src="data:image/foo;base64,0123456789">
</body>

</html>
