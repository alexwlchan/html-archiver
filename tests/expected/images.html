<html>

<head>
  <meta charset="utf-8">
  <title>This is a page with images</title>
</head>

<body>
  <p>This is a page to test how html-archiver handles images.</p>

  <p>Because <code>&gt;img&lt;</code> tags can refer to external resources, they should be inlined where possible.</p>

  <!-- A basic image that works -->
  <img src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAEYAAABGCAIAAAD+THXTAAAAAXNSR0IArs4c6QAAAAlwSFlzAAALEwAACxMBAJqcGAAAAjBpVFh0WE1MOmNvbS5hZG9iZS54bXAAAAAAADx4OnhtcG1ldGEgeG1sbnM6eD0iYWRvYmU6bnM6bWV0YS8iIHg6eG1wdGs9IlhNUCBDb3JlIDUuNC4wIj4KICAgPHJkZjpSREYgeG1sbnM6cmRmPSJodHRwOi8vd3d3LnczLm9yZy8xOTk5LzAyLzIyLXJkZi1zeW50YXgtbnMjIj4KICAgICAgPHJkZjpEZXNjcmlwdGlvbiByZGY6YWJvdXQ9IiIKICAgICAgICAgICAgeG1sbnM6eG1wPSJodHRwOi8vbnMuYWRvYmUuY29tL3hhcC8xLjAvIgogICAgICAgICAgICB4bWxuczp0aWZmPSJodHRwOi8vbnMuYWRvYmUuY29tL3RpZmYvMS4wLyI+CiAgICAgICAgIDx4bXA6Q3JlYXRvclRvb2w+QWNvcm4gdmVyc2lvbiA0LjUuODwveG1wOkNyZWF0b3JUb29sPgogICAgICAgICA8dGlmZjpDb21wcmVzc2lvbj41PC90aWZmOkNvbXByZXNzaW9uPgogICAgICAgICA8dGlmZjpZUmVzb2x1dGlvbj43MjwvdGlmZjpZUmVzb2x1dGlvbj4KICAgICAgICAgPHRpZmY6WFJlc29sdXRpb24+NzI8L3RpZmY6WFJlc29sdXRpb24+CiAgICAgIDwvcmRmOkRlc2NyaXB0aW9uPgogICA8L3JkZjpSREY+CjwveDp4bXBtZXRhPgpDyBsAAAAA5ElEQVRoBe2YQQ7DIAwETZWHpy9v+4WVQJOuJmfMxjNGQqzPtH2vtoZmbOkfnGpJSwgBBw/BHoYWWrrmDiE8fnmhJVt6/NT9flBLWkIIOHgI9jC00NL61D0+FFqypfCkIsu1hGAPQ7UUAkOWawnBHoau8fYQIgOWe5YA6HGklmJkQIGWAOhxpJZiZEDBNe+264ODB8xRHKmlGBlQoCUAehyppRgZUFBoaU3d40OhJVsCTnscqaUYGVCgJQB6HKmlGBlQcN1A6NlIB+8s3z27a2kPx7O7aOks3z27a2kPx7O7FFr6AhBlCArDDUhZAAAAAElFTkSuQmCC"/>

  <!-- We can include it more than once -->
  <img src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAEYAAABGCAIAAAD+THXTAAAAAXNSR0IArs4c6QAAAAlwSFlzAAALEwAACxMBAJqcGAAAAjBpVFh0WE1MOmNvbS5hZG9iZS54bXAAAAAAADx4OnhtcG1ldGEgeG1sbnM6eD0iYWRvYmU6bnM6bWV0YS8iIHg6eG1wdGs9IlhNUCBDb3JlIDUuNC4wIj4KICAgPHJkZjpSREYgeG1sbnM6cmRmPSJodHRwOi8vd3d3LnczLm9yZy8xOTk5LzAyLzIyLXJkZi1zeW50YXgtbnMjIj4KICAgICAgPHJkZjpEZXNjcmlwdGlvbiByZGY6YWJvdXQ9IiIKICAgICAgICAgICAgeG1sbnM6eG1wPSJodHRwOi8vbnMuYWRvYmUuY29tL3hhcC8xLjAvIgogICAgICAgICAgICB4bWxuczp0aWZmPSJodHRwOi8vbnMuYWRvYmUuY29tL3RpZmYvMS4wLyI+CiAgICAgICAgIDx4bXA6Q3JlYXRvclRvb2w+QWNvcm4gdmVyc2lvbiA0LjUuODwveG1wOkNyZWF0b3JUb29sPgogICAgICAgICA8dGlmZjpDb21wcmVzc2lvbj41PC90aWZmOkNvbXByZXNzaW9uPgogICAgICAgICA8dGlmZjpZUmVzb2x1dGlvbj43MjwvdGlmZjpZUmVzb2x1dGlvbj4KICAgICAgICAgPHRpZmY6WFJlc29sdXRpb24+NzI8L3RpZmY6WFJlc29sdXRpb24+CiAgICAgIDwvcmRmOkRlc2NyaXB0aW9uPgogICA8L3JkZjpSREY+CjwveDp4bXBtZXRhPgpDyBsAAAAA5ElEQVRoBe2YQQ7DIAwETZWHpy9v+4WVQJOuJmfMxjNGQqzPtH2vtoZmbOkfnGpJSwgBBw/BHoYWWrrmDiE8fnmhJVt6/NT9flBLWkIIOHgI9jC00NL61D0+FFqypfCkIsu1hGAPQ7UUAkOWawnBHoau8fYQIgOWe5YA6HGklmJkQIGWAOhxpJZiZEDBNe+264ODB8xRHKmlGBlQoCUAehyppRgZUFBoaU3d40OhJVsCTnscqaUYGVCgJQB6HKmlGBlQcN1A6NlIB+8s3z27a2kPx7O7aOks3z27a2kPx7O7FFr6AhBlCArDDUhZAAAAAElFTkSuQmCC"/>

  <!-- An image that is missing -->
  <img src="missing.jpg">

  <!-- Test that other attributes are preserved -->
  <img alt="a green snake with a red tongue" src="data:image/gif;base64,R0lGODdhRgBGAJEAAAAAAP8AAACAQP///yH5BAQAAAAALAAAAABGAEYAAALKnI+py+0Po5y02ouz3rz7D4biSJbmiaaLwLbuC8fy/GL0jeeurff+fvkJZYAioMUbKlnGIyu5HDaRigAkupw+E9Yr9ssKiMcSMLgaMX8XXQdNpXnDobA5vWa3yPOUPb8889cXKAgoU2gYg5imFnXX6PMIqSM5iVNp6TeYGRnE2an32YMpihdaern4oKmqwNqK8AprIDtbC3vbmqu6u9iL+OuKmno6TFhsrOiZfLjMXOf8DIQsvUVdTWqcPTzb7f0NHi4+Tl5ufo5RAAA7" style="border: 1px solid black;">

  <!-- Test that URL encoded junk is interpreted correctly -->
  <img src="data:image/jpeg;base64,/9j/4AAQSkZJRgABAQAASABIAAD/4QCURXhpZgAATU0AKgAAAAgABAEaAAUAAAABAAAAPgEbAAUAAAABAAAARgExAAIAAAAUAAAATodpAAQAAAABAAAAYgAAAAAAAABIAAAAAQAAAEgAAAABQWNvcm4gdmVyc2lvbiA0LjUuOAAAA6ABAAMAAAABAAEAAKACAAQAAAABAAAARqADAAQAAAABAAAARgAAAAD/7QA4UGhvdG9zaG9wIDMuMAA4QklNBAQAAAAAAAA4QklNBCUAAAAAABDUHYzZjwCyBOmACZjs+EJ+/8AAEQgARgBGAwEiAAIRAQMRAf/EAB8AAAEFAQEBAQEBAAAAAAAAAAABAgMEBQYHCAkKC//EALUQAAIBAwMCBAMFBQQEAAABfQECAwAEEQUSITFBBhNRYQcicRQygZGhCCNCscEVUtHwJDNicoIJChYXGBkaJSYnKCkqNDU2Nzg5OkNERUZHSElKU1RVVldYWVpjZGVmZ2hpanN0dXZ3eHl6g4SFhoeIiYqSk5SVlpeYmZqio6Slpqeoqaqys7S1tre4ubrCw8TFxsfIycrS09TV1tfY2drh4uPk5ebn6Onq8fLz9PX29/j5+v/EAB8BAAMBAQEBAQEBAQEAAAAAAAABAgMEBQYHCAkKC//EALURAAIBAgQEAwQHBQQEAAECdwABAgMRBAUhMQYSQVEHYXETIjKBCBRCkaGxwQkjM1LwFWJy0QoWJDThJfEXGBkaJicoKSo1Njc4OTpDREVGR0hJSlNUVVZXWFlaY2RlZmdoaWpzdHV2d3h5eoKDhIWGh4iJipKTlJWWl5iZmqKjpKWmp6ipqrKztLW2t7i5usLDxMXGx8jJytLT1NXW19jZ2uLj5OXm5+jp6vLz9PX29/j5+v/bAEMAAQEBAQEBAgEBAgMCAgIDBAMDAwMEBQQEBAQEBQYFBQUFBQUGBgYGBgYGBgcHBwcHBwgICAgICQkJCQkJCQkJCf/bAEMBAQEBAgICBAICBAkGBQYJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCf/dAAQABf/aAAwDAQACEQMRAD8A/wA/+iv9fr/iFy/4IUf9EM/8ubxP/wDLaj/iFy/4IUf9EM/8ubxP/wDLagD/ACBaK/1+v+IXL/ghR/0Qz/y5vE//AMtqP+IXL/ghR/0Qz/y5vE//AMtqAP8AIFor/X6/4hcv+CFH/RDP/Lm8T/8Ay2o/4hcv+CFH/RDP/Lm8T/8Ay2oA/wAgWiv9fr/iFy/4IUf9EM/8ubxP/wDLav5Av+DrX/glx+wn/wAE1/8AhQv/AAxT4G/4Qv8A4TT/AISn+2f+Jnqmo/af7O/sj7L/AMhK7uvL8v7VN/q9m7f827auAD+QKiiigD//0I/+I53/AKtd/wDL2/8AvDR/xHO/9Wu/+Xt/94a/gDr9f/2W/wDggr/wVh/bR+BOhftL/s0fCn/hJfBPiX7V/Zupf27oVn532O6ls5/3F7qEFwmy4gkT541zt3LlSCQD+n3/AIjnf+rXf/L2/wDvDR/xHO/9Wu/+Xt/94a/AH/iFy/4Lr/8ARDP/AC5vDH/y2o/4hcv+C6//AEQz/wAubwx/8tqAP3+/4jnf+rXf/L2/+8NH/Ec7/wBWu/8Al7f/AHhr8Af+IXL/AILr/wDRDP8Ay5vDH/y2r8AaAP8Ab5/4Jcftz/8ADyj9hTwN+2r/AMIt/wAIX/wmn9p/8Sb7d/aP2b+ztTu9N/4+vs9r5nmfZfM/1Kbd+3nbuP8AIF/wfO/82u/9zt/7gK/f7/g1x/5QUfAz/uZv/Un1avwB/wCD53/m13/udv8A3AUAfwB0UUUAf//R/wA/+v8AX6/4Ncf+UFHwM/7mb/1J9Wr/ACBa/wBfr/g1x/5QUfAz/uZv/Un1agD9/qK/gD/4Pnf+bXf+52/9wFfwB0Af7/Ff4A9FFAH+v1/wa4/8oKPgZ/3M3/qT6tX4A/8AB87/AM2u/wDc7f8AuAr9/v8Ag1x/5QUfAz/uZv8A1J9Wr8Af+D53/m13/udv/cBQB/AHRRRQB//S/wA/+v8AX6/4Ncf+UFHwM/7mb/1J9Wr/ACBa/r9/4Jcf8HWv/Dtf9hTwN+xV/wAKF/4TT/hC/wC0/wDic/8ACU/2d9p/tHU7vUv+PX+yLry/L+1eX/rn3bN3G7aAD+/z9uf/AIJcfsJ/8FKP+EW/4bW8Df8ACaf8IX9u/sb/AImeqad9m/tH7P8Aav8AkG3dr5nmfZYf9Zv27Pl27mz8Af8AELl/wQo/6IZ/5c3if/5bV+AP/Ec7/wBWu/8Al7f/AHho/wCI53/q13/y9v8A7w0Afv8Af8QuX/BCj/ohn/lzeJ//AJbUf8QuX/BCj/ohn/lzeJ//AJbV+AP/ABHO/wDVrv8A5e3/AN4aP+I53/q13/y9v/vDQB/b7+y3+y58Cf2LvgToX7NH7NGhf8I14J8Nfav7N037VdXnk/bLqW8n/f3ss9w++4nkf55Gxu2rhQAP4gv+D53/AJtd/wC52/8AcBR/xHO/9Wu/+Xt/94a/AH/gud/wXO/4fR/8Ku/4td/wrX/hWv8Abf8AzG/7Y+2/2z9g/wCnCx8nyfsP/TTf5n8O35gD8AaKKKAP/9P/AD/6KKKACiiigAooooAKKKKACiiigD//2Q==">

  <!-- Test that if an image is already base64-encoded, it's ignored -->
  <img src="data:image/foo;base64,0123456789">
</body>

</html>
